File: backend/app/schemas/billing.py
"""Pydantic schemas for billing and Stripe integration."""

from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, ConfigDict, Field, EmailStr

from app.models.tenant import SubscriptionStatus


class SignupInitiate(BaseModel):
    """Schema for initiating signup before Stripe checkout."""
    tenant_name: str = Field(
        ...,
        min_length=3,
        max_length=50,
        pattern=r'^[a-z0-9][a-z0-9-]*[a-z0-9]$',
        description="URL-safe tenant slug (lowercase letters, numbers, hyphens)"
    )
    email: EmailStr = Field(..., description="Admin email for the tenant")
    plan_id: int = Field(..., description="Subscription plan ID")
    display_name: Optional[str] = Field(None, max_length=255, description="Human-readable tenant name")
    billing_period: str = Field(
        default="monthly",
        description="Billing period: 'monthly' or 'yearly'"
    )


class SignupInitiateResponse(BaseModel):
    """Response after initiating signup."""
    checkout_url: str = Field(..., description="Stripe Checkout URL to redirect user to")
    session_id: str = Field(..., description="Stripe Checkout session ID")


class CheckoutSessionCreate(BaseModel):
    """Schema for creating a checkout session for existing tenant (upgrade/change plan)."""
    plan_id: int
    success_url: Optional[str] = None
    cancel_url: Optional[str] = None


class BillingPortalResponse(BaseModel):
    """Response for billing portal redirect."""
    url: str = Field(..., description="Stripe Billing Portal URL")


class SubscriptionInfo(BaseModel):
    """Current subscription information."""
    status: SubscriptionStatus
    plan_name: str
    plan_id: int
    current_period_end: Optional[datetime] = None
    trial_ends_at: Optional[datetime] = None
    grace_period_ends_at: Optional[datetime] = None
    cancel_at_period_end: bool = False
    
    model_config = ConfigDict(from_attributes=True, use_enum_values=True)


class UsageInfo(BaseModel):
    """Current usage information for the tenant."""
    users_count: int
    users_limit: int
    users_remaining: int
    studies_count: int
    studies_limit: int
    studies_remaining: int
    tracker_items_count: int
    tracker_items_limit: int
    tracker_items_remaining: int


class BillingOverview(BaseModel):
    """Complete billing overview for tenant dashboard."""
    subscription: SubscriptionInfo
    usage: UsageInfo
    can_add_users: bool
    can_add_studies: bool
    upgrade_available: bool
    has_billing_account: bool = Field(
        False,
        description="Whether tenant has a Stripe billing account (customer ID and subscription)"
    )


class PlanInfo(BaseModel):
    """Public plan information for pricing page."""
    id: int
    name: str
    display_name: str
    description: Optional[str] = None
    price_monthly: int = Field(..., description="Price in cents")
    price_yearly: Optional[int] = Field(None, description="Yearly price in cents (if available)")
    max_users: int
    max_studies: int
    max_tracker_items: int
    features: Optional[dict] = None
    is_popular: bool = False

    model_config = ConfigDict(from_attributes=True)


class PlanListResponse(BaseModel):
    """List of available plans."""
    plans: List[PlanInfo]


class WebhookEvent(BaseModel):
    """Stripe webhook event (for logging/debugging)."""
    event_id: str
    event_type: str
    processed_at: datetime
    tenant_id: Optional[int] = None
    success: bool
    error_message: Optional[str] = None


class AutoRenewToggle(BaseModel):
    """Request to toggle auto-renewal for subscription."""
    cancel_auto_renew: bool = Field(
        ...,
        description="True to cancel auto-renewal (subscription ends at period end), False to resume auto-renewal"
    )


class AutoRenewResponse(BaseModel):
    """Response after toggling auto-renewal."""
    cancel_at_period_end: bool = Field(..., description="Whether subscription will cancel at period end")
    current_period_end: Optional[datetime] = Field(None, description="End date of current billing period")
    message: str = Field(..., description="Human-readable status message")

File: react-frontend/src/api/endpoints/billing.ts
/**
 * Billing API endpoints for subscription management
 */

import { apiClient } from '../client';

// Types
export interface Plan {
  id: number;
  name: string;
  display_name: string;
  description: string | null;
  price_monthly: number;
  price_yearly: number | null;
  max_users: number;
  max_studies: number;
  max_tracker_items: number;
  features: Record<string, boolean> | null;
  is_popular: boolean;
}

export interface PlansResponse {
  plans: Plan[];
}

export interface SignupInitiateRequest {
  tenant_name: string;
  email: string;
  plan_id: number;
  display_name?: string;
  billing_period?: 'monthly' | 'yearly';
}

export interface SignupInitiateResponse {
  checkout_url: string;
  session_id: string;
}

export interface SubscriptionInfo {
  status: 'trialing' | 'active' | 'past_due' | 'canceled' | 'unpaid';
  plan_name: string;
  plan_id: number;
  current_period_end: string | null;
  trial_ends_at: string | null;
  grace_period_ends_at: string | null;
  cancel_at_period_end: boolean;
}

export interface UsageInfo {
  users_count: number;
  users_limit: number;
  users_remaining: number;
  studies_count: number;
  studies_limit: number;
  studies_remaining: number;
  tracker_items_count: number;
  tracker_items_limit: number;
  tracker_items_remaining: number;
}

export interface BillingOverview {
  subscription: SubscriptionInfo;
  usage: UsageInfo;
  can_add_users: boolean;
  can_add_studies: boolean;
  upgrade_available: boolean;
  has_billing_account: boolean;
}

export interface BillingPortalResponse {
  url: string;
}

export interface BillingStatus {
  stripe_configured: boolean;
  trial_days: number;
  grace_period_days: number;
}

export interface AutoRenewRequest {
  cancel_auto_renew: boolean;
}

export interface AutoRenewResponse {
  cancel_at_period_end: boolean;
  current_period_end: string | null;
  message: string;
}

// API Functions

/**
 * Get list of available subscription plans (public)
 */
export const getPlans = async (): Promise<PlansResponse> => {
  const response = await apiClient.get<PlansResponse>('/api/v1/billing/plans');
  return response.data;
};

/**
 * Initiate signup - creates Stripe Checkout session (public)
 */
export const initiateSignup = async (data: SignupInitiateRequest): Promise<SignupInitiateResponse> => {
  const response = await apiClient.post<SignupInitiateResponse>('/api/v1/billing/signup/initiate', data);
  return response.data;
};

/**
 * Get billing overview for current tenant (authenticated)
 */
export const getBillingOverview = async (): Promise<BillingOverview> => {
  const response = await apiClient.get<BillingOverview>('/api/v1/billing/overview');
  return response.data;
};

/**
 * Create Stripe Billing Portal session (admin only)
 */
export const createBillingPortal = async (): Promise<BillingPortalResponse> => {
  const response = await apiClient.post<BillingPortalResponse>('/api/v1/billing/portal');
  return response.data;
};

/**
 * Check if Stripe is configured (public)
 */
export const getBillingStatus = async (): Promise<BillingStatus> => {
  const response = await apiClient.get<BillingStatus>('/api/v1/billing/status');
  return response.data;
};

/**
 * Toggle auto-renewal for subscription (admin only)
 */
export const toggleAutoRenew = async (data: AutoRenewRequest): Promise<AutoRenewResponse> => {
  const response = await apiClient.post<AutoRenewResponse>('/api/v1/billing/auto-renew', data);
  return response.data;
};

// Helper functions

/**
 * Format price in cents to display string
 */
export const formatPrice = (cents: number): string => {
  if (cents === 0) return 'Custom';
  return `$${(cents / 100).toFixed(0)}`;
};

/**
 * Get subscription status display info
 */
export const getStatusDisplay = (status: SubscriptionInfo['status']): { label: string; color: string } => {
  switch (status) {
    case 'active':
      return { label: 'Active', color: 'text-green-600' };
    case 'trialing':
      return { label: 'Trial', color: 'text-blue-600' };
    case 'past_due':
      return { label: 'Payment Due', color: 'text-yellow-600' };
    case 'canceled':
      return { label: 'Canceled', color: 'text-gray-500' };
    case 'unpaid':
      return { label: 'Unpaid', color: 'text-red-600' };
    default:
      return { label: status, color: 'text-gray-600' };
  }
};
